src/js/media-controller.js

```diff
@@ -450,101 +428,132 @@ const setAttr = (child, attrName, attrValue) => {
+  // We (currently) don't check if descendants of the `media` (e.g. <video/>) are Media State Receivers
+  // See also: `propagateMediaState`
+  if (isMediaSlotElementDescendant(rootNode)) {
+    return;
+  }
 
-const toNextMediaUIElementsState = (mutationsList = []) => {
-  const mediaChromeNodesState = mutationsList.reduce((prevMediaChromeNodesState, mutationRecord) => {
-    const { addedNodes = [], removedNodes = [], type, target, attributeName } = mutationRecord;
-    if (type === 'childList') {
-      const addedMediaChromeNodes = Array.prototype.flatMap.call(addedNodes, getMediaUIElementDescendants);
-      const removedMediaChromeNodes = Array.prototype.flatMap.call(removedNodes, getMediaUIElementDescendants);
-      prevMediaChromeNodesState.addedNodes.push(...addedMediaChromeNodes);
-      prevMediaChromeNodesState.removedNodes.push(...removedMediaChromeNodes);
-      return prevMediaChromeNodesState;
-    }
-    /** option 1, controller side */
-    if (type === 'attributes' && attributeName === MediaUIAttributes.MEDIA_CHROME_ATTRIBUTES) {
-      const attrs = getMediaUIAttributesFrom(target);
-      if (attrs.length) {
-        prevMediaChromeNodesState.addedNodes.push(target);
-      }
-      else {
-        prevMediaChromeNodesState.removedNodes.push(target);
-      }
-      return prevMediaChromeNodesState;
+  const traverseForMediaStateReceiversSync = (rootNode, mediaStateReceiverCallback) => {
+    // The rootNode is itself a Media State Receiver
+    if (isMediaStateReceiver(rootNode)) {
+      mediaStateReceiverCallback(rootNode);
     }
-    return prevMediaChromeNodesState;
-  }, {
-    addedNodes: [],
-    removedNodes: [],
-  });
-  return mediaChromeNodesState;
+
+    const { children = [] } = rootNode ?? {};
+    const shadowChildren = rootNode?.shadowRoot?.children ?? [];
+    const allChildren = [...children, ...shadowChildren];
+
+    // Traverse all children (including shadowRoot children) to see if they are/have Media State Receivers
+    allChildren.forEach(child => traverseForMediaStateReceivers(child, mediaStateReceiverCallback));
+  };
+
+  // Custom Elements (and *only* Custom Elements) must have a hyphen ("-") in their name. So, if the rootNode is
+  // a custom element (aka has a hyphen in its name), wait until it's defined before attempting traversal to determine
+  // whether or not it or its descendants are Media State Receivers.
+  // IMPORTANT NOTE: We're intentionally *always* waiting for the `whenDefined()` Promise to resolve here
+  // (instead of using `window.customElements.get(name)` to check if a custom element is already defined/registered)
+  // because we encountered some reliability issues with the custom element instances not being fully "ready", even if/when
+  // they are available in the registry via `window.customElements.get(name)`.
+  const name = rootNode?.nodeName.toLowerCase();
+  if (name.includes('-') && !isMediaStateReceiver(rootNode)) {
```

Comment:
Took me a minute to figure out why we check for `!isMediaStateReceiver` rather than just treat every undefined custom element the same, but I get now the assumption is that if it's a state receiver then it must be defined. Which isn't technically true with the special attr that could exist before it's defined. I can't come up with why someone would ever do that so I'll leave it up to you if it's worth addressing.